src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.cs

```diff
@@ -1,69 +1,263 @@
                                                                              customTags: WellKnownDiagnosticTags.Telemetry);
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(DefaultRule, PropertyRule, MultipleRule, VisibilityRule);
 
         public override void Initialize(AnalysisContext analysisContext)
-        { 
-            
+        {
+            analysisContext.RegisterSymbolAction(DoAnalysis, SymbolKind.Method);
+        }
+
+        private static void DoAnalysis(SymbolAnalysisContext symbolContext)
+        {
+            var methodSymbol = (IMethodSymbol)symbolContext.Symbol;
+            var location = methodSymbol.OriginalDefinition.Locations.First();
+            var typeSymbol = methodSymbol.ContainingSymbol as ITypeSymbol;
+            var operatorName = methodSymbol.Name;
+            if (typeSymbol != null && (methodSymbol.MethodKind == MethodKind.UserDefinedOperator || methodSymbol.MethodKind == MethodKind.Conversion))
+            {
+                if (IsPropertyExpected(operatorName) && operatorName != OpFalseText)
+                {
+                    // don't report a diagnostic on the `op_False` method because then the user would see two diagnostics for what is really one error
+                    // special-case looking for `IsTrue` instance property
+                    // named properties can't be overloaded so there will only ever be 0 or 1
+                    var property = typeSymbol.GetMembers(IsTrueText).OfType<IPropertySymbol>().SingleOrDefault();
+                    if (property == null || property.Type.SpecialType != SpecialType.System_Boolean)
+                    {
+                        symbolContext.ReportDiagnostic(CreateDiagnostic(PropertyRule, location, AddAlternateText, IsTrueText, operatorName));
+                    }
+                    else if (!property.IsPublic())
+                    {
+                        symbolContext.ReportDiagnostic(CreateDiagnostic(VisibilityRule, property.OriginalDefinition.Locations.First(), FixVisibilityText, IsTrueText, operatorName));
+                    }
+                }
+                else
+                {
+                    var expectedGroup = GetExpectedAlternateMethodGroup(operatorName, methodSymbol.ReturnType);
+                    if (expectedGroup == null)
+                    {
+                        // no alternate methods required
+                        return;
+                    }
+
+                    var matchedMethods = new List<IMethodSymbol>();
+                    var unmatchedMethods = new HashSet<string>() { expectedGroup.AlternateMethod1 };
+                    if (expectedGroup.AlternateMethod2 != null)
+                    {
+                        unmatchedMethods.Add(expectedGroup.AlternateMethod2);
+                    }
+
+                    foreach (var candidateMethod in typeSymbol.GetMembers().OfType<IMethodSymbol>())
```

Comment:
I had your same concern, but I couldn't think of a nice way to cache the method symbols and I didn't want to go the other route of analyzing all types then looking for operator overloads, since the `if` block at the top of this method will filter out a lot of non-operator-overload symbols.